### PR TITLE
daf_persistence: make Butler.datasetExists work with composite datasets

### PR DESCRIPTION
**Pull request by n8pease**

### CODE REVIEW

File: python/lsst/daf/persistence/butler.py [hunk 1]
```diff
@@ -1245,17 +1245,23 @@ def datasetExists(self, datasetType, dataId={}, **rest):
...
-        for repoData in self._repos.inputs():
-            if not dataId.tag or len(dataId.tag.intersection(repoData.tags)) > 0:
-                location = repoData.repo.map(datasetType, dataId)
-                if location and location.repository.exists(location):
-                    break
+        exists = None
+        if isinstance(location, ButlerComposite):
+            for name, componentInfo in location.componentInfo.items():
+                if componentInfo.subset:
+                    subset = self.subset(datasetType=componentInfo.datasetType, dataId=location.dataId)
```
Comment (kennywlo) [inline python/lsst/daf/persistence/butler.py]: PEP8:  line above is too long - consider moving dataId... to next line.

File: python/lsst/daf/persistence/butler.py [hunk 2]
```diff
@@ -1245,17 +1245,23 @@ def datasetExists(self, datasetType, dataId={}, **rest):
...
+                    exists = all([obj.datasetExists() for obj in subset])
                 else:
-                    location = None
-
-        return bool(location)
+                    exists = self.datasetExists(componentInfo.datasetType, location.dataId)
```
Comment (kennywlo) [inline python/lsst/daf/persistence/butler.py]: PEP8: line above too long.

Comment (RobertLuptonTheGood) [inline python/lsst/daf/persistence/butler.py]: The indentation looks odd.  I believe that we are sticking to the 110 line limit; how long is this one?

File: python/lsst/daf/persistence/butler.py [hunk 3]
```diff
@@ -1245,17 +1245,23 @@ def datasetExists(self, datasetType, dataId={}, **rest):
...
+            return False
 
-        location = None
-        for repoData in self._repos.inputs():
-            if not dataId.tag or len(dataId.tag.intersection(repoData.tags)) > 0:
-                location = repoData.repo.map(datasetType, dataId)
-                if location and location.repository.exists(location):
-                    break
+        exists = None
+        if isinstance(location, ButlerComposite):
```
Comment (kennywlo) [inline python/lsst/daf/persistence/butler.py]: I find ambiguity in the name "ButlerComposite" and its relation to location in the comparison, especially the first part "Butler", in the local context - a comment would help.

File: python/lsst/daf/persistence/butler.py [hunk 4]
```diff
@@ -1245,17 +1245,23 @@ def datasetExists(self, datasetType, dataId={}, **rest):
...
+        if location is None:
+            return False
 
-        location = None
-        for repoData in self._repos.inputs():
-            if not dataId.tag or len(dataId.tag.intersection(repoData.tags)) > 0:
-                location = repoData.repo.map(datasetType, dataId)
-                if location and location.repository.exists(location):
-                    break
+        exists = None
```
Comment (kennywlo) [inline python/lsst/daf/persistence/butler.py]: suggestion: exists = False (given exists being boolean)

File: python/lsst/daf/persistence/butler.py (hunk 1 above)
Comment (n8pease) [inline python/lsst/daf/persistence/butler.py]: we allow 110, this is 104.

File: python/lsst/daf/persistence/butler.py (hunk 2 above)
Comment (n8pease) [inline python/lsst/daf/persistence/butler.py]: @RobertLuptonTheGood are you reading the code on the Conversation tab? ...Don't compare the red indentation to the green indentation. If you look at this line under View Changes you'll see the indentation is correct.
@kennywlo the line is 92 columns, < 110.

File: python/lsst/daf/persistence/butler.py [hunk 5]
```diff
@@ -1245,17 +1245,23 @@ def datasetExists(self, datasetType, dataId={}, **rest):
...
+            for name, componentInfo in location.componentInfo.items():
+                if componentInfo.subset:
+                    subset = self.subset(datasetType=componentInfo.datasetType, dataId=location.dataId)
+                    exists = all([obj.datasetExists() for obj in subset])
                 else:
-                    location = None
-
-        return bool(location)
+                    exists = self.datasetExists(componentInfo.datasetType, location.dataId)
+                if exists is False:
```
Comment (n8pease) [inline python/lsst/daf/persistence/butler.py]: I've been shot down for that before; "the double negative is hard to parse"

**Comment by n8pease** [comment]
@kennywlo Please re-review the changes under the latest commit.
